src/core/simulation.jl

```diff
@@ -333,3 +364,161 @@ function build!(sim::Simulation; kwargs...)
+    ini_cond_key::ICKey,
+    chronology::IntraStageChronology,
+    sim::Simulation,
+)
+    ini_cond_vector = get_initial_conditions(stage.internal.psi_container)[ini_cond_key]
+    for ic in get_initial_conditions(stage.internal.psi_container)[ini_cond_key]
+        name = device_name(ic)
+        interval_chronology = get_stage_interval_chronology(sim, stage.name)
+        var_value =
+            get_stage_variable(interval_chronology, (stage => stage), name, ic.update_ref)
+        cache = get_cache(stage, ic.cache_key)
+        quantity = calculate_ic_quantity(ini_cond_key, ic, var_value, cache)
+        PJ.fix(ic.value, quantity)
+    end
+    return
+end
+
+""" Updates the initial conditions of the stage"""
+function initial_condition_update!(
+    stage::Stage,
+    ini_cond_key::ICKey,
+    chronology::InterStageChronology,
+    sim::Simulation,
+)
+    execution_index = get_execution_order(sim)
+    ini_cond_vector = get_initial_conditions(stage.internal.psi_container)[ini_cond_key]
+    for ic in ini_cond_vector
+        name = device_name(ic)
+        current_ix = get_current_execution_index(sim)
+        source_stage_ix = current_ix == 1 ? length(execution_index) : current_ix - 1
+        source_stage = get_stage(sim, execution_index[source_stage_ix])
+        source_stage_name = get_stage_name(sim, source_stage)
+        interval_chronology = get_stage_interval_chronology(sim.sequence, source_stage_name)
+        var_value = get_stage_variable(
+            interval_chronology,
+            (source_stage => stage),
+            name,
+            ic.update_ref,
+        )
+        if isnothing(ic.cache_type)
+            cache = nothing
+        else
+            cache = get_cache(sim, ic.cache_type, ini_cond_key.device_type)
+        end
+        quantity = calculate_ic_quantity(ini_cond_key, ic, var_value, cache)
+        PJ.fix(ic.value, quantity)
+    end
+
+    return
+end
+
+function _update_caches!(sim::Simulation, stage::Stage)
+    for cache in stage.internal.caches
+        update_cache!(sim, cache, stage)
+    end
+    return
+end
+
+################################Cache Update################################################
+# TODO: Need to be careful here if 2 stages modify the same cache. This function might need
+# dispatch on the Statge{OpModel} to assing different actions. e.g. HAUC and DAUC
+function update_cache!(
+    sim::Simulation,
+    key::CacheKey{TimeStatusChange, D},
+    stage::Stage,
+) where {D <: PSY.Device}
+    c = get_cache(sim, key)
+    variable = get_variable(stage.internal.psi_container, c.ref)
+
+    for t in 1:get_end_of_interval_step(stage), name in variable.axes[1]
+        device_status = JuMP.value(variable[name, t])
+        if c.value[name][:status] == device_status
+            c.value[name][:count] += 1.0
+        elseif c.value[name][:status] != device_status
+            c.value[name][:count] = 1.0
+            c.value[name][:status] = device_status
+        end
+    end
+
```

Comment:
It would probably be a good idea to log or record these actions.

Reply:
I will add a debug statement and we should implement a better logging solution later.